Mage/src/main/java/mage/players/Player.java

```diff
@@ -105,7 +106,11 @@ default boolean isComputer() {
 
     void addAbility(Ability ability);
 
-    Counters getCounters();
+    /**
+     * The counters should be manipulated (adding or removing counters) with the appropriate addCounters/removeCounters/getCounterCount methods.
+     * This returns a copy for specific usage, to make sure the Player's counters are not altered from there.
+     */
+    Counters getCopyCounters();
```

Comment:
Must keep old method `getCounters()`. If you need to use shared code then use `getCountersForModification()` or like that. It will help to auto-document all modification use cases. All that additional methods and protection are overdeveloped -- concurrent error will catch wrong usage immediately.